src/test/CharterManager-unit.t.sol

```diff
@@ -0,0 +1,644 @@
+        return manager.proxy(address(this));
+    }
+    function gems() public view returns (uint256) {
+        return VatHelpersLike(address(adapter.vat())).gem(adapter.ilk(), proxy());
+    }
+    function urn() public view returns (uint256, uint256) {
+        return VatHelpersLike(address(adapter.vat())).urns(adapter.ilk(), proxy());
+    }
+    function dai() public view returns (uint256) {
+        return VatHelpersLike(address(adapter.vat())).dai(address(this));
+    }
+    function allow(address usr) public {
+        manager.allow(address(usr));
+    }
+    function disallow(address usr) public {
+        manager.disallow(address(usr));
+    }
+    function frob(int256 dink, int256 dart) public {
+        manager.frob(address(adapter), address(this), address(this), address(this), dink, dart);
+    }
+    function frob(address u, address v, address w, int256 dink, int256 dart) public {
+        manager.frob(address(adapter), u, v, w, dink, dart);
+    }
+    function frobDirect(address u, address v, address w, int256 dink, int256 dart) public {
+        VatLike(manager.vat()).frob(adapter.ilk(), u, v, w, dink, dart);
+    }
+    function flux(address src, address dst, uint256 wad) public {
+        manager.flux(address(adapter), src, dst, wad);
+    }
+    function fluxDirect(address src, address dst, uint256 wad) public {
+        VatLike(manager.vat()).flux(adapter.ilk(), src, dst, wad);
+    }
+    function quit() public {
+        manager.quit(adapter.ilk(), address(this));
+    }
+    function quit(address dst) public {
+        manager.quit(adapter.ilk(), dst);
+    }
+    function hope(address vat, address usr) public {
+        Vat(vat).hope(usr);
+    }
+    function nope(address vat, address usr) public {
+        Vat(vat).nope(usr);
+    }
+
+    function try_call(address addr, bytes calldata data) external returns (bool) {
+        bytes memory _data = data;
+        assembly {
+            let ok := call(gas(), addr, 0, add(_data, 0x20), mload(_data), 0, 0)
+            let free := mload(0x40)
+            mstore(free, ok)
+            mstore(0x40, add(free, 32))
+            revert(free, 32)
+        }
+    }
+    function can_call(address addr, bytes memory data) internal returns (bool) {
+        (bool ok, bytes memory success) = address(this).call(
+            abi.encodeWithSignature(
+                "try_call(address,bytes)"
+            , addr
+            , data
+            ));
+        ok = abi.decode(success, (bool));
+        if (ok) return true;
+    }
+    function can_exit(address usr, uint256 val) public returns (bool) {
+        bytes memory call = abi.encodeWithSignature
+        ("exit(address,address,uint256)", address(adapter), usr, val);
+        return can_call(address(manager), call);
+    }
+}
+
+contract CharterManagerTest is TestBase {
+
+    Token               gem;
+    Token               dai;
+    Vat                 vat;
+    Vow                 vow;
+    Jug                 jug;
```

Comment:
You can get rid of `jug` completely if you take my advice below to set rates directly via `Vat.fold`.

Reply:
Done.